src/firefly/js/visualize/XYPlot.jsx

```diff
@@ -34,288 +35,610 @@ const plotParamsShape = PropTypes.shape({
-                                onHighlightChange(highlighted);
+                            var highlightedIdx = this.rowIdx ? this.rowIdx : this.series.data.indexOf(this);
+                            if (highlightedIdx !== highlighted.rowIdx) {
+                                onHighlightChange(highlightedIdx);
                             }
                         }
                     }
                 }
             };
 
+            if (!decimateKey) {
+                let pushFunc;
+                if (selectInfo) {
+                    const selectInfoCls = SelectInfo.newInstance(selectInfo, 0);
+                    // split data into selected and unselected
+                    pushFunc = (numdata, nrow, idx) => {
+                        selectInfoCls.isSelected(idx) ?
+                            numdata.selected.push({x: nrow[0], y: nrow[1], rowIdx: nrow[2]}) :
+                            numdata.unselected.push({x: nrow[0], y: nrow[1], rowIdx: nrow[2]});
+                    };
+                } else {
+                    pushFunc = (numdata, nrow) => {
+                        numdata.unselected.push({x: nrow[0], y: nrow[1], rowIdx: nrow[2]});
+                    };
+                }
+                const numericData = rows.reduce((numdata, arow, idx) => {
+                    const nrow = arow.map(toNumber);
+                    pushFunc(numdata, nrow, idx);
+                    return numdata;
+                }, {selected: [], unselected: []});
+
+                marker = {symbol: 'circle'};
+                allSeries = [
+                    {
+                        id: UNSELECTED,
+                        name: UNSELECTED,
+                        color: unselectedColor,
+                        data: numericData.unselected,
+                        marker,
+                        turboThreshold: 0,
+                        point
+                    },
+                    {
+                        id: SELECTED,
+                        name: SELECTED,
+                        color: selectedColor,
+                        data: numericData.selected,
+                        marker,
+                        turboThreshold: 0,
+                        point
+                    }
+                ];
+            } else {
+                const {xUnitPx, yUnitPx} = this.getDeciSymbolSize(chart, decimateKey);
+                marker = {symbol: 'rectangle', radius: xUnitPx/2, hD: (xUnitPx-yUnitPx)/2};
+
+                const {xMin, xUnit, yMin, yUnit} = parseDecimateKey(decimateKey);
+                const getCenter = (xval,yval) => {
+                    return {
+                        // bitwise operators convert operands to 32-bit integer
+                        // hence they can be used as a fast way to truncate a float to an integer
+                        x: xMin+(~~((xval-xMin)/xUnit)+0.5)*xUnit,
+                        y: yMin+(~~((yval-yMin)/yUnit)+0.5)*yUnit
+                    };
```

Comment:
I am glad you documented this! I have seen it before but would not have remembered. If you are worried about the performance it might be better to use a for loop instead of reduce on line 430. I would normally say reduce is better but I have heard in performance critical javascript that a for loop is faster.